File: packages/igloo-ts/src/widgets/container.ts
import { Container as WitContainer } from 'iced:app/container@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { Padding } from 'iced:app/padding@0.1.0';
import type { Horizontal, Vertical } from 'iced:app/alignment@0.1.0';
import type { Pixels } from 'iced:app/shared@0.1.0';
import { Element, toElement, type ElementLike, type IntoElement } from '../element.js';

/**
 * Builder for creating Container widgets.
 * A Container wraps a single child with optional padding, sizing, and alignment.
 *
 * @example
 * ```typescript
 * const centered = Container.new(Text.new('Centered')).center(Length.fill());
 * ```
 *
 * @typeParam Msg - The message type of the wrapped content, inferred from `new`.
 */
export class Container<Msg = never> implements IntoElement<Msg> {
  private raw: WitContainer;

  private constructor(content: ElementLike<Msg>) {
    this.raw = new WitContainer(toElement(content).inner);
  }

  /** Create a new Container builder with the given content element */
  static new<const Msg = never>(content: ElementLike<Msg>): Container<Msg> {
    return new Container(content);
  }

  /** Set the padding around the content */
  padding(padding: Padding): this {
    this.raw.padding(padding);
    return this;
  }

  /** Set the container width */
  width(width: Length): this {
    this.raw.width(width);
    return this;
  }

  /** Set the container height */
  height(height: Length): this {
    this.raw.height(height);
    return this;
  }

  /** Set the maximum width in pixels */
  maxWidth(maxWidth: Pixels): this {
    this.raw.maxWidth(maxWidth);
    return this;
  }

  /** Set the maximum height in pixels */
  maxHeight(maxHeight: Pixels): this {
    this.raw.maxHeight(maxHeight);
    return this;
  }

  /** Center the content horizontally with the given width */
  centerX(width: Length): this {
    this.raw.centerX(width);
    return this;
  }

  /** Center the content vertically with the given height */
  centerY(height: Length): this {
    this.raw.centerY(height);
    return this;
  }

  /** Center the content both horizontally and vertically */
  center(size: Length): this {
    this.raw.center(size);
    return this;
  }

  /** Align content to the left with the given width */
  alignLeft(width: Length): this {
    this.raw.alignLeft(width);
    return this;
  }

  /** Align content to the right with the given width */
  alignRight(width: Length): this {
    this.raw.alignRight(width);
    return this;
  }

  /** Align content to the top with the given height */
  alignTop(height: Length): this {
    this.raw.alignTop(height);
    return this;
  }

  /** Align content to the bottom with the given height */
  alignBottom(height: Length): this {
    this.raw.alignBottom(height);
    return this;
  }

  /** Set horizontal alignment */
  alignX(align: Horizontal): this {
    this.raw.alignX(align);
    return this;
  }

  /** Set vertical alignment */
  alignY(align: Vertical): this {
    this.raw.alignY(align);
    return this;
  }

  /** Enable or disable clipping of content */
  clip(clip: boolean = true): this {
    this.raw.clip(clip);
    return this;
  }

  /** Convert to Element */
  intoElement(): Element<Msg> {
    return new Element(WitContainer.intoElement(this.raw));
  }
}

File: packages/igloo-py/src/igloo_py/widgets/container.py
"""Container widget builder."""

from typing import TYPE_CHECKING

from wit_world.imports.container import Container as WitContainer

from ..element import Element, ElementLike, IntoElement, to_element

if TYPE_CHECKING:
    from ..types.alignment import WitHorizontal, WitVertical
    from ..types.length import WitLength
    from ..types.padding import WitPadding


class Container(IntoElement):
    """
    Builder for creating Container widgets.
    A Container is a widget that wraps a single child with optional padding,
    sizing, and alignment.

    Example:
        centered = Container.new(Text.new("Centered")).center(Length.fill())
    """

    def __init__(self, content: ElementLike) -> None:
        self._raw = WitContainer(to_element(content).inner)

    @classmethod
    def new(cls, content: ElementLike) -> "Container":
        """Create a new Container builder with the given content element."""
        return cls(content)

    def padding(self, padding: "WitPadding") -> "Container":
        """Set the padding around the content."""
        self._raw.padding(padding)
        return self

    def width(self, width: "WitLength") -> "Container":
        """Set the container width."""
        self._raw.width(width)
        return self

    def height(self, height: "WitLength") -> "Container":
        """Set the container height."""
        self._raw.height(height)
        return self

    def max_width(self, max_width: float) -> "Container":
        """Set the maximum width in pixels."""
        self._raw.max_width(max_width)
        return self

    def max_height(self, max_height: float) -> "Container":
        """Set the maximum height in pixels."""
        self._raw.max_height(max_height)
        return self

    def center_x(self, width: "WitLength") -> "Container":
        """Center the content horizontally with the given width."""
        self._raw.center_x(width)
        return self

    def center_y(self, height: "WitLength") -> "Container":
        """Center the content vertically with the given height."""
        self._raw.center_y(height)
        return self

    def center(self, size: "WitLength") -> "Container":
        """Center the content both horizontally and vertically."""
        self._raw.center(size)
        return self

    def align_left(self, width: "WitLength") -> "Container":
        """Align content to the left with the given width."""
        self._raw.align_left(width)
        return self

    def align_right(self, width: "WitLength") -> "Container":
        """Align content to the right with the given width."""
        self._raw.align_right(width)
        return self

    def align_top(self, height: "WitLength") -> "Container":
        """Align content to the top with the given height."""
        self._raw.align_top(height)
        return self

    def align_bottom(self, height: "WitLength") -> "Container":
        """Align content to the bottom with the given height."""
        self._raw.align_bottom(height)
        return self

    def align_x(self, align: "WitHorizontal") -> "Container":
        """Set horizontal alignment."""
        self._raw.align_x(align)
        return self

    def align_y(self, align: "WitVertical") -> "Container":
        """Set vertical alignment."""
        self._raw.align_y(align)
        return self

    def clip(self, clip: bool = True) -> "Container":
        """Enable or disable clipping of content."""
        self._raw.clip(clip)
        return self

    def into_element(self) -> Element:
        """Convert to Element."""
        return Element(WitContainer.into_element(self._raw))
